packages/sdk/client-services/src/packlets/tests/text.test.ts

```diff
@@ -0,0 +1,210 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import * as fc from 'fast-check';
+import { ModelRunSetup } from 'fast-check';
+import waitForExpect from 'wait-for-expect';
+
+import { Client, Document, Text } from '@dxos/client';
+import { PublicKey } from '@dxos/keys';
+import { afterTest, describe, test } from '@dxos/test';
+import { Doc } from '@dxos/text-model';
+import { ComplexMap, ComplexSet, range } from '@dxos/util';
+
+import { joinCommonSpace, TestBuilder } from '../testing';
+
+type Model = {
+  text: string;
+  peers: ComplexSet<PublicKey>;
+};
+
+type Real = {
+  spaceKey: PublicKey;
+  peers: ComplexMap<PublicKey, Client>;
+};
+
+const assertState = async (model: Model, real: Real) => {
+  await waitForExpect(() => {
```

Comment:
is it possible to make this event driven? e.g., test based on throttled update/join events?
i don't immediately see how, but waitFor always makes me uneasy.

Reply:
this is copied from the network manager property tests, unsure if it's needed yet, we'll see. Wasn't sure if we'd have to wait for replication or not and how easy that would be to do with events in these fuzzy tests